### PR TITLE
Obg 35: fix DKB  Adapter 

### PR DESCRIPTION
**Pull request by clovisgakam**

### CODE REVIEW

File: opba-protocols/xs2a-protocol/src/main/resources/processes/xs2a/authorization/oauth2/xs2a-oauth2-step.bpmn20.xml [hunk 1]
```diff
@@ -23,6 +23,15 @@
       <conditionExpression xsi:type="tFormalExpression">#{xs2aConsentInfo.isOauth2TokenAvailableAndReadyToUse(CONTEXT)}</conditionExpression>
     </sequenceFlow>
     <sequenceFlow id="sid-cf198380-6fed-408b-8641-0a918a74a2db" sourceRef="oauth2ExchangeCodeToToken" targetRef="sid-e0dbf285-c6c2-4571-ac5e-9038fcbf8448"/>
+    <sequenceFlow id="sid-39bca444-9c5a-4432-9662-e41c0b2e1231" name="Embedded pre Auth required" sourceRef="checkThatOauth2TokenIsPresentAndNotExpired" targetRef="xs2aAskForPassword">
+      <conditionExpression xsi:type="tFormalExpression">#{xs2aConsentInfo.isEmbeddedPreAuthNeeded(CONTEXT)}</conditionExpression>
+    </sequenceFlow>
+    <serviceTask id="embeddedOauth2ExchangeCodeToToken" name="Get Embedded Access Token" flowable:async="true" flowable:exclusive="true" flowable:delegateExpression="${xs2aEmbeddedPreAuthorization}" flowable:triggerable="false" isForCompensation="false"/>
```
Comment (valb3r) [inline opba-protocols/xs2a-protocol/src/main/resources/processes/xs2a/authorization/oauth2/xs2a-oauth2-step.bpmn20.xml:29]: This element name obviously is not `Get Embedded Access Token`, but rather `Exchange password (embedded pre-step) to token`

File: opba-protocols/xs2a-protocol/src/main/resources/processes/xs2a/authorization/oauth2/xs2a-oauth2-step.bpmn20.xml [hunk 2]
```diff
@@ -23,6 +23,15 @@
...
+    <sequenceFlow id="sid-1b0f3e86-4b13-450c-9f6f-e0975dddedc8" sourceRef="embeddedOauth2ExchangeCodeToToken" targetRef="sid-e0dbf285-c6c2-4571-ac5e-9038fcbf8448"/>
+    <serviceTask id="xs2aAskForPassword" name="Get Embedded Access Token" isForCompensation="false" flowable:async="true" flowable:exclusive="true" flowable:delegateExpression="${xs2aAskForPassword}" flowable:triggerable="true"/>
```
Comment (valb3r) [inline opba-protocols/xs2a-protocol/src/main/resources/processes/xs2a/authorization/oauth2/xs2a-oauth2-step.bpmn20.xml:31]: This element name obviously is not `Get Embedded Access Token`, but rather `Ask PSU for password for embedded token`

File: opba-protocols/xs2a-protocol/src/main/resources/processes/xs2a/authorization/oauth2/xs2a-oauth2-step.bpmn20.xml [hunk 3]
```diff
@@ -23,6 +23,15 @@
       <conditionExpression xsi:type="tFormalExpression">#{xs2aConsentInfo.isOauth2TokenAvailableAndReadyToUse(CONTEXT)}</conditionExpression>
     </sequenceFlow>
     <sequenceFlow id="sid-cf198380-6fed-408b-8641-0a918a74a2db" sourceRef="oauth2ExchangeCodeToToken" targetRef="sid-e0dbf285-c6c2-4571-ac5e-9038fcbf8448"/>
+    <sequenceFlow id="sid-39bca444-9c5a-4432-9662-e41c0b2e1231" name="Embedded pre Auth required" sourceRef="checkThatOauth2TokenIsPresentAndNotExpired" targetRef="xs2aAskForPassword">
```
Comment (valb3r) [inline opba-protocols/xs2a-protocol/src/main/resources/processes/xs2a/authorization/oauth2/xs2a-oauth2-step.bpmn20.xml:26]: The `Embedded pre Auth required` is rather `PSU password required for embedded OAuth2 token`

File: opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/oauth2/Xs2aEmbeddedPreAuthorization.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        ValidatedQueryHeaders<Xs2aOauth2WithCodeParameters, Xs2aOauth2Headers> validated = extractor.forExecution(context);
+
+        logResolver.log("getToken with parameters: {}", validated.getQuery(), validated.getHeaders());
+        EmbeddedPreAuthorisationRequest embeddedPreAuthorisationRequest = new EmbeddedPreAuthorisationRequest();
+        embeddedPreAuthorisationRequest.setPassword(context.getPsuPassword());
+        embeddedPreAuthorisationRequest.setUsername(context.getPsuId());
+        TokenResponse response = this.embeddedPreAuthorisationService.getToken(embeddedPreAuthorisationRequest, validated.getHeaders().toHeaders());
+        if (response.getTokenType() == null) {
+            response.setTokenType("Bearer");
```
Comment (valb3r) [inline opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/oauth2/Xs2aEmbeddedPreAuthorization.java:44]: `"Bearer"` is a constant literal

File: opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/oauth2/Xs2aEmbeddedPreAuthorization.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    @Override
+    @SneakyThrows
+    protected void doRealExecution(DelegateExecution execution, Xs2aContext context) {
+        logResolver.log("doRealExecution: execution ({}) with context ({})", execution, context);
+
+        ValidatedQueryHeaders<Xs2aOauth2WithCodeParameters, Xs2aOauth2Headers> validated = extractor.forExecution(context);
+
+        logResolver.log("getToken with parameters: {}", validated.getQuery(), validated.getHeaders());
+        EmbeddedPreAuthorisationRequest embeddedPreAuthorisationRequest = new EmbeddedPreAuthorisationRequest();
```
Comment (valb3r) [inline opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/oauth2/Xs2aEmbeddedPreAuthorization.java:39]: `EmbeddedPreAuthorisationRequest` should be created and mapped with `Xs2aEmbeddedPreAuthorization.Extractor` and not manually

File: opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/oauth2/Xs2aEmbeddedPreAuthorization.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        embeddedPreAuthorisationRequest.setPassword(context.getPsuPassword());
+        embeddedPreAuthorisationRequest.setUsername(context.getPsuId());
+        TokenResponse response = this.embeddedPreAuthorisationService.getToken(embeddedPreAuthorisationRequest, validated.getHeaders().toHeaders());
+        if (response.getTokenType() == null) {
+            response.setTokenType("Bearer");
+        }
+        logResolver.log("getToken response: {}", response);
```
Comment (valb3r) [inline opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/oauth2/Xs2aEmbeddedPreAuthorization.java:46]: Please double check that token is not logged here

File: opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/authenticate/embedded/Xs2aAisAuthenticateUserConsentWithPin.java [hunk 7]
```diff
@@ -50,7 +50,9 @@ protected void doValidate(DelegateExecution execution, Xs2aContext context) {
     @Override
     protected void doRealExecution(DelegateExecution execution, Xs2aContext context) {
         logResolver.log("doRealExecution: execution ({}) with context ({})", execution, context);
-
+        if (context.isEmbeddedPreAuthDone()) {
```
Comment (valb3r) [inline opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/authenticate/embedded/Xs2aAisAuthenticateUserConsentWithPin.java:53]: This check should be removed from here and instead one should use it in diagram exclusive gateway

**Review by valb3r — CHANGES_REQUESTED** [review]

File: opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/oauth2/Xs2aEmbeddedPreAuthorization.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            );
+
+
+    }
+
+    @Override
+    protected void doValidate(DelegateExecution execution, Xs2aContext context) {
+        logResolver.log("doValidate: execution ({}) with context ({})", execution, context);
+
+     //   validator.validate(execution, context, this.getClass(), extractor.forValidation(context));
```
Comment (valb3r) [inline opba-protocols/xs2a-protocol/src/main/java/de/adorsys/opba/protocol/xs2a/service/xs2a/oauth2/Xs2aEmbeddedPreAuthorization.java:66]: Also, according to https://github.com/adorsys/open-banking-gateway/pull/1284/files#r682317167
suggest to request PSU password/login here

File: opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml [hunk 9]
```diff
@@ -120,4 +120,10 @@
         </sql>
     </changeSet>
 
+    <changeSet id="set_preffered_approach_dkb" author="mhr" context="xs2a-banks">
+        <sql dbms="postgresql" stripComments="true">
+            UPDATE opb_bank_profile SET preferred_approach = 'EMBEDDED', try_to_use_preferred_approach = true WHERE adapter_id = 'dkb-adapter';
```
Comment (valb3r) [inline opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml:125]: Please introduce a special approach for DKB or special approach setting that indicates we should execute OAuth2-embedded pre-step

File: opba-protocols/xs2a-protocol-tests/xs2a-bdd-wiremock/src/main/java/de/adorsys/opba/protocol/xs2a/tests/e2e/wiremock/mocks/MockServers.java [hunk 10]
```diff
@@ -263,18 +271,23 @@ public SELF embedded_mock_of_sandbox_for_max_musterman_transactions_running_for_
 
 
     @SneakyThrows
-    private void startWireMock(WireMockConfiguration config) {
+    private  void startWireMock(WireMockConfiguration config) {
         sandbox = new WireMockServer(config);
         sandbox.start();
         BankProfile bankProfile = bankProfileJpaRepository.findByBankUuid("53c47f54-b9a4-465a-8f77-bc6cd5f0cf46").get();
```
[the remark this answers is not in the document]
Comment (clovisgakam) [reply opba-protocols/xs2a-protocol-tests/xs2a-bdd-wiremock/src/main/java/de/adorsys/opba/protocol/xs2a/tests/e2e/wiremock/mocks/MockServers.java:277]: Done in  startWireMockForDkb method

File: opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml (hunk 9 above)
Comment (clovisgakam) [reply opba-db/src/main/resources/migration/migrations/0000-create-table-banks.xml]: I am not sure if a new setting is required as it is actually done base on the bank profile  preferred_approach  field and the PreAuthorisationException thrown by the adapter

**Comment by valb3r** [comment]
Please use https://github.com/adorsys/open-banking-gateway/tree/OBG-35-dkb-updated instead

**Review by valb3r — APPROVED** [review]

**Pull request merged**